### PR TITLE
Optimize discovery module.

### PR DESCRIPTION
**Pull request by GuoYL123**
Optimize  discovery module:
1. support AZ affinity
  Instances of different AZ will take a long time. The same AZ instance is preferred.
2. use revision reduce request:
   Access the service center with revision, and return 304 when there is no updated information.
3. handle empty server list 
   After the service center goes online again after a period of downtime, because there is no heartbeat renewal, the service center will clean up the normal instances and return them to the empty list.

### CODE REVIEW

File: spring-cloud-huawei-common/src/main/java/com/huaweicloud/common/transport/Response.java [hunk 1]
```diff
@@ -28,6 +32,21 @@
 
   private String content;
 
+  private Map<String, String> headers = new HashMap<>();
+
+  public String getHeader(String key) {
+    if (!headers.containsKey(key)) {
```
Comment (liubao68) [inline spring-cloud-huawei-common/src/main/java/com/huaweicloud/common/transport/Response.java:38]: This check seams not neccessary.

File: spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandler.java [hunk 2]
```diff
@@ -39,9 +41,15 @@
 
   private static List<ServiceInstance> instanceList = Collections.emptyList();
 
+  public static Map<String, String> serviceRevision = new ConcurrentHashMap<>();
```
Comment (liubao68) [inline spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandler.java:44]: Define a setter is better, e.g. addServiceReivison(String, String). Or must difine it to be final .

File: spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/client/ServiceCombClientTest.java [hunk 3]
```diff
@@ -210,7 +210,7 @@ public void getInstances(@Injectable
       }
     };
     ServiceCombClient serviceCombClient = new ServiceCombClient(url, httpTransport);
-    List<ServiceInstance> actual = serviceCombClient.getInstances(microservice);
+    List<ServiceInstance> actual = serviceCombClient.getInstances(microservice,null);
```
Comment (liubao68) [inline spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/client/ServiceCombClientTest.java:213]: Formate this code

File: spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandlerTest.java [hunk 4]
```diff
@@ -54,7 +54,7 @@ public void getInstances(@Injectable ServiceCombClient serviceCombClient)
     microservice.setServiceName("testservice");
     new Expectations() {
       {
-        serviceCombClient.getInstances(microservice);
+        serviceCombClient.getInstances(microservice,"0");
```
Comment (liubao68) [inline spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandlerTest.java:57]: Format this code

File: spring-cloud-huawei-common/src/main/java/com/huaweicloud/common/transport/Response.java (hunk 1 above)
Comment (GuoYL123) [reply spring-cloud-huawei-common/src/main/java/com/huaweicloud/common/transport/Response.java]: done

File: spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandler.java (hunk 2 above)
Comment (GuoYL123) [reply spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandler.java]: change it to final

File: spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/client/ServiceCombClientTest.java (hunk 3 above)
Comment (GuoYL123) [reply spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/client/ServiceCombClientTest.java]: done

File: spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandlerTest.java (hunk 4 above)
Comment (GuoYL123) [reply spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceHandlerTest.java]: done

File: spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/ribbon/ServiceCombServerListTest.java [hunk 5]
```diff
@@ -85,21 +92,23 @@ public URI getUri() {
...
+        return map;
       }
     };
     instanceList.add(serviceInstance);
     new Expectations() {
       {
         iClientConfig.getClientName();
         result = "serviceid11";
-        serviceCombClient.getInstances((Microservice) any);
+        serviceCombClient.getInstances((Microservice) any,null);
```
Comment (liubao68) [inline spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/ribbon/ServiceCombServerListTest.java:105]: format code

File: spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/KieAddrSeekerImpl.java [hunk 6]
```diff
@@ -28,7 +28,7 @@ public String getKieAddr() {
     microservice.setServiceName("servicecomb-kie");
     List<ServiceInstance> instanceList;
     try {
-      instanceList = serviceCombClient.getInstances(microservice);
+      instanceList = serviceCombClient.getInstances(microservice,null);
```
Comment (liubao68) [inline spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/KieAddrSeekerImpl.java:31]: format code

File: spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceCache.java [hunk 7]
```diff
@@ -35,19 +35,17 @@ public static void initService(List<Microservice> list) {
...
     }
 
     public static void initInsList(List<MicroserviceInstance> list, String serviceName) {
         list.forEach(ins -> {
             ins.setServiceName(serviceName);
-            ins.getEndpoints().forEach(ep -> {
-                microserviceList.put(ep.replaceAll("rest://", ""), ins);
-            });
+            ins.getEndpoints()
+                .forEach(ep -> microserviceList.put(ep.replaceAll("rest://", ""), ins));
```
Comment (liubao68) [inline spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceCache.java:48]: I am not sure where `getMicroserviceIns` is called. Seams use microservice instance id as the key is better. Not all endpoints starts with rest.

Comment (GuoYL123) [reply spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/MicroserviceCache.java:48]: Ribbon's  com.netflix.loadbalancer.Server choose the endpint as the instance id, so I use  endpoint.    
In spring cloud all endpoints with rest.

File: spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/ribbon/ServiceCombServerListTest.java (hunk 5 above)
Comment (GuoYL123) [reply spring-cloud-huawei-servicecomb-discovery/src/test/java/com/huaweicloud/servicecomb/discovery/ribbon/ServiceCombServerListTest.java]: done

File: spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/KieAddrSeekerImpl.java (hunk 6 above)
Comment (GuoYL123) [reply spring-cloud-huawei-servicecomb-discovery/src/main/java/com/huaweicloud/servicecomb/discovery/discovery/KieAddrSeekerImpl.java]: done

**Pull request merged**